component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -0,0 +1,199 @@
+import org.wso2.carbon.identity.oauth2.client.authentication.AbstractOAuthClientAuthenticator;
+import org.wso2.carbon.identity.oauth2.client.authentication.OAuthClientAuthnException;
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls.utils.MutualTLSUtil;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import javax.servlet.http.HttpServletRequest;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls.utils.MutualTLSUtil.JAVAX_SERVLET_REQUEST_CERTIFICATE;
+
+/**
+ * This class is responsible for authenticating OAuth clients with Mutual TLS. The client will present
+ * client certificate presented to the authorization server during TLS handshake. As a result of successful
+ * validation of the certificate at web container, the certificate will be available in request attributes. This
+ * authenticator will authenticate the client by matching the certificate presented during handshake against the
+ * certificate registered for the client.
+ */
+public class MutualTLSClientAuthenticator extends AbstractOAuthClientAuthenticator {
+
+    private static Log log = LogFactory.getLog(MutualTLSClientAuthenticator.class);
+
+    /**
+     * @param request                 HttpServletRequest which is the incoming request.
+     * @param bodyParams              Body parameter map of the request.
+     * @param oAuthClientAuthnContext OAuth client authentication context.
+     * @return Whether the authentication is successful or not.
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public boolean authenticateClient(HttpServletRequest request, Map<String, List> bodyParams,
+                                      OAuthClientAuthnContext oAuthClientAuthnContext)
+            throws OAuthClientAuthnException {
+
+        // In case if the client ID is not set from canAuthenticate method.
+        if (StringUtils.isEmpty(oAuthClientAuthnContext.getClientId())) {
+            oAuthClientAuthnContext.setClientId(getClientId(request, bodyParams, oAuthClientAuthnContext));
+        }
+
+        try {
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticating client : " + oAuthClientAuthnContext.getClientId() + " with public " +
+                        "certificate.");
+            }
+
+            X509Certificate requestCert;
+            Object certObject = request.getAttribute(JAVAX_SERVLET_REQUEST_CERTIFICATE);
+            if (certObject instanceof X509Certificate[]) {
+                X509Certificate[] cert = (X509Certificate[]) certObject;
+                requestCert = cert[0];
+            } else if (certObject instanceof X509Certificate){
+                requestCert = (X509Certificate) certObject;
+            } else {
+                return false;
```

Comment:
Shall we print a debug lot to show that we didn't get a proper certificate ?. Otherwise it's really confusing when authentication fails without any reason.